httpcore/_async/http2.py

```diff
@@ -91,20 +102,24 @@ async def request(
     ) -> Tuple[bytes, int, bytes, List[Tuple[bytes, bytes]], AsyncByteStream]:
         timeout = {} if timeout is None else timeout
 
-        async with self.init_lock:
-            if not self.sent_connection_init:
-                # The very first stream is responsible for initiating the connection.
-                self.state = ConnectionState.ACTIVE
-                await self.send_connection_init(timeout)
-                self.sent_connection_init = True
-
-            try:
-                stream_id = self.h2_state.get_next_available_stream_id()
-            except NoAvailableStreamIDError:
-                self.state = ConnectionState.FULL
-                raise NewConnectionRequired()
-            else:
-                self.state = ConnectionState.ACTIVE
+        await self.streams_semaphore.acquire()
+        try:
+            async with self.init_lock:
+                if not self.sent_connection_init:
+                    # The very first stream is responsible for initiating the connection.
+                    self.state = ConnectionState.ACTIVE
+                    await self.send_connection_init(timeout)
+                    self.sent_connection_init = True
+
+                try:
+                    stream_id = self.h2_state.get_next_available_stream_id()
+                except NoAvailableStreamIDError:
+                    self.state = ConnectionState.FULL
+                    raise NewConnectionRequired()
+                else:
+                    self.state = ConnectionState.ACTIVE
+        finally:
+            self.streams_semaphore.release()
```

Comment:
Okay, so I *think* the wrapping here isn't quite correct. It'd make sense to instead wrap up the portion of this method *after* the connection init, so...

```python
# Wrap this up inside the stream semaphore
h2_stream = AsyncHTTP2Stream(stream_id=stream_id, connection=self)
self.streams[stream_id] = h2_stream
self.events[stream_id] = []
return await h2_stream.request(method, url, headers, stream, timeout)
```

Since that'd ensure that we hold the semaphore for the duration of a request/response.

In fact, it's actually a bit more awkward that tho. We *don't* want to release the semaphore until the client has *closed* the response we've returned. So really we want to release the semaphore either on an exception occuring within this block, or on the response being closed.

Reply:
So, something like...

```python
async def request(...):
    ...
    await self.streams_semaphore.acquire()
    try:
        h2_stream = AsyncHTTP2Stream(stream_id=stream_id, connection=self)
        self.streams[stream_id] = h2_stream
        self.events[stream_id] = []
        return await h2_stream.request(method, url, headers, stream, timeout)
    except:
        self.streams_semaphore.release()
        raise

...
async def close_stream(self, stream_id: int) -> None:
    self.streams_semaphore.release()
    ...
```

Alternately we could pass the semaphore to `AsyncHTTP2Stream`, and wrap up the behaviour there instead. (That approach might feel a bit more tightly constrained)